ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java

```diff
@@ -201,120 +189,42 @@ private void runReAct(
-        tmpParameters.put("prompt.tool_response", promptToolResponse);
-
-        StringSubstitutor promptSubstitutor = new StringSubstitutor(tmpParameters, "${parameters.", "}");
-        prompt = promptSubstitutor.replace(prompt);
-
-        final List<String> inputTools = new ArrayList<>();
-        for (Map.Entry<String, Tool> entry : tools.entrySet()) {
-            String toolName = Optional.ofNullable(entry.getValue().getName()).orElse(entry.getValue().getType());
-            inputTools.add(toolName);
-        }
-
-        prompt = AgentUtils.addPrefixSuffixToPrompt(parameters, prompt);
-        prompt = AgentUtils.addToolsToPrompt(tools, parameters, inputTools, prompt);
-        prompt = AgentUtils.addIndicesToPrompt(parameters, prompt);
-        prompt = AgentUtils.addExamplesToPrompt(parameters, prompt);
-        prompt = AgentUtils.addChatHistoryToPrompt(parameters, prompt);
-        prompt = AgentUtils.addContextToPrompt(parameters, prompt);
+        boolean verbose = parameters.containsKey(VERBOSE) && Boolean.parseBoolean(parameters.get(VERBOSE));
+        boolean traceDisabled = parameters.containsKey(DISABLE_TRACE) && Boolean.parseBoolean(parameters.get(DISABLE_TRACE));
 
+        Map<String, String> tmpParameters = constructLLMParams(llm, parameters);
+        String prompt = constructLLMPrompt(tools, parameters, inputTools, tmpParameters);
         tmpParameters.put(PROMPT, prompt);
 
-        List<ModelTensors> modelTensors = new ArrayList<>();
-
-        List<ModelTensors> cotModelTensors = new ArrayList<>();
-        cotModelTensors
-            .add(
-                ModelTensors
-                    .builder()
-                    .mlModelTensors(
-                        List
-                            .of(
-                                ModelTensor.builder().name(MLAgentExecutor.MEMORY_ID).result(sessionId).build(),
-                                ModelTensor.builder().name(MLAgentExecutor.PARENT_INTERACTION_ID).result(parentInteractionId).build()
-                            )
-                    )
-                    .build()
-            );
+        List<ModelTensors> traceTensors = createModelTensors(sessionId, parentInteractionId);
 
         StringBuilder scratchpadBuilder = new StringBuilder();
-        StringSubstitutor tmpSubstitutor = new StringSubstitutor(
-            ImmutableMap.of(SCRATCHPAD, scratchpadBuilder.toString()),
-            "${parameters.",
-            "}"
-        );
+        StringSubstitutor tmpSubstitutor = new StringSubstitutor(Map.of(SCRATCHPAD, scratchpadBuilder.toString()), "${parameters.", "}");
         AtomicReference<String> newPrompt = new AtomicReference<>(tmpSubstitutor.replace(prompt));
         tmpParameters.put(PROMPT, newPrompt.get());
-
-        String maxIteration = Optional.ofNullable(tmpParameters.get("max_iteration")).orElse("3");
+        String finalPrompt = prompt;
```

Comment:
I am a little confused by this finalPrompt, it looks like we will discard the old scratchpad content? Let's have a discussion offline.